### ISSUE TITLE
kubectl UX - bad error

### ISSUE THREAD

**Issue by thockin** [GoogleCloudPlatform/kubernetes#8609]
$ kubectl create -f /tmp/pod.yaml 
error: no objects passed to create

There's SOMETHING in that file - I am not sure what's wrong with it.

**Comment by ArtfulCoder** [comment]
Hidden character(s) ?

**Comment by philips** [comment]
Seems very much related to #8610 :)

**Comment by thockin** [comment]
The bug was that I forgot a -, thereby turning a list into a map, which
failed SOMEWHERE, causing the above error.

**Comment by apatil** [comment]
I just got this by putting booleans and integers with no quotes in a yaml environment section, eg:

```yaml
...
env:
  - name: SMTP_PORT
    value: 587
...
```

Fixed it by replacing `587` with `"587"`, but to figure that out I had to step through the env lines with emacs, deleting them and then attempting `kubectl create`. :)

**Comment by matthiasr** [comment]
Here's a workaround to find out what it's choking on: read the YAML from stdin. With a borked file, I get:

```
% kubectl create -f prometheus-rc.yaml
error: no objects passed to create
```
but
```
kubectl create -f - <prometheus-rc.yaml
W0607 18:25:00.309319   15695 visitor.go:474] Could not read an encoded object from STDIN: unable to load "STDIN": json: cannot unmarshal object into Go value of type []v1beta3.ContainerPort
error: no objects passed to create
```
This will give you an idea where the error lies.
Funnily, even though it says JSON, it does totally accept YAML, but there must be some subtle difference between reading from stdin or from file.

**Comment by arun-gupta** [comment]
Try:
```
kubectl --v=5 create -f file.yaml
```

It might give you some more detailed information.

**Comment by mulloymorrow** [comment]
```bash
I0612 11:14:53.705129    1811 visitor.go:223] Unable to load file "some-replication-controller.yaml": unable to load "some-replication-controller.yaml": json: cannot unmarshal number into Go value of type string
```

**Comment by mikedanese** [comment]
@mulloymorrow show me some-replication-controller.yaml! put quotes on your strings!

**Comment by matthiasr** [comment]
I guess this is a completely separate issue, but would it be possible to coerce numbers into strings where that makes sense ( = label values )?

**Comment by mulloymorrow** [comment]
@mikedanese , yeah it works when we have quotes on strings. Just a little misleading to the end user. The exception `error: no objects passed to create` could be a bit more informative.

**Comment by mikedanese** [comment]
@mulloymorrow I think this is fixed by #9357. Can you try upgrading your kubectl? A few errors in the resource visitor loop were being logged at invisible log levels and are being printed to stderr now.

**Comment by mikedanese** [comment]
@matschaffer It's a bit tricky. It's actually a _"feature"_ of the golang stdlib "encoding/json" package. See:

http://play.golang.org/p/sdR_gXKDpH

I'm not sure if it's possible to fix while still using "encoding/json". File an issue if you want to see it.
